### PR TITLE
remove hdb_views for inserts

### PR DESCRIPTION
**Pull request by paf31**
### Description

This is a precursor to #3533. It is an early WIP, but I'd like to put up this PR for review. In particular, I still need to:

- [x] investigate whether I can avoid passing around as many aliases for CTEs
- [x] add more documentation
- [x] fix up the tests and write new tests
- [x] test the CLI changes
- [x] (if we decide it's the right approach), add the `err(text)` function to `hdb_views` or `hdb_catalog` and create a migration for it

### Affected components

- [x] Server

### Related Issues
#3710

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [ ] No
- [x] Yes
  - [x] Updated docs with SQL for downgrading the catalog 

#### Metadata

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes
  - Does `run_sql` auto manages the new metadata through schema diffing?
    - [ ] Yes
    - [ ] Not required
  - Does `run_sql` auto manages the definitions of metadata on renaming?
    - [ ] Yes
    - [ ] Not required
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?
    - [ ] Yes
    - [ ] Not required

#### GraphQL
- [ ] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated:
   - [ ] New types and typenames are correlated
   
   

#### Breaking changes

- [x] No Breaking changes
- [ ] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [ ] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     

  2. GraphQL API

     Schema Generation:
     
     
     - [ ] Change in any `NamedType`
     - [ ] Change in table field names
     

     Schema Resolve:-
     
     - [ ] Change in treatment of `null` value for any input fields 
     

  3. Logging

     - [ ] Log `JSON` schema has changed
     - [ ] Log `type` names have changed

### CODE REVIEW

**Comment by paf31** [comment]
Tests are now passing 🎉

**Comment by 0x777** [comment]
@paf31 I looked through the changes, from what I understand, the generated insert statement is of this form:

```sql
WITH 
  "<table>__mutation_result_alias" AS ( insert statement returning * ),
  "<table>__inserted" AS ( SELECT .. ),
  "<table>__checks" AS ( SELECT bool_and( .. WHEN check_expression ..) ) 
SELECT
  "<table>__inserted".*
FROM
  "<table>__inserted",
  "<table>__checks"
```

I experimented with a `check` condition which involves a condition on another table through a relationship, in such cases the following insert statement has a better plan (check expression is computed in returning):
```sql
WITH 
  "<table>__mutation_result_alias" AS 
    ( insert statement returning *, check_expression AS _check_expression_value ),
  "<table>__inserted" AS ( SELECT .. ),
  "<table>__checks" AS ( SELECT bool_and( .. WHEN _check_expression_value ..) ) 
SELECT
  "<table>__inserted".*
FROM
  "<table>__inserted",
  "<table>__checks"
```

The two statements that I used are as follows:
<details><summary>Statement 1</summary>

```sql
explain analyze with "address_mutation" as (
  update
    address a
  set
    location = 'blr'
  where
    id = 1 returning a.*
),
"address_check" as (
  select
    bool_and(
      case
        when (
          exists (
            select
              1
            from
              "user" u
            where
              u.id = a.user_id
          )
        ) THEN NULL
        ELSE "hdb_catalog"."check_violation"()
      end
    )
  from
    "address_mutation" a
),
"address_return" as (
  select
    json_agg(a.*) as returning
  from
    "address_mutation" a
)
select
  a.*
from
  "address_return" a,
  "address_check";
```
</details>


<details><summary>Statement 2</summary>

```sql
explain analyze with "address_mutation" as (
  update
    address a
  set
    location = 'blr'
  where
    id = 1 returning a.*,
    (
      exists (
        select
          1
        from
          "user" u
        where
          u.id = a.user_id
      )
    ) as _check_condition
),
"address_check" as (
  select
    bool_and(
      case
        when (a._check_condition) THEN NULL
        else "hdb_catalog"."check_violation"()
      end
    )
  from
    "address_mutation" a
),
"address_return" as (
  select
    json_agg(a.*) as returning
  from
    "address_mutation" a
)
select
  a.*
from
  "address_return" a,
  "address_check";
```
</details>

Regarding the `check_violation` function defined in `hdb_catalog`, it would be best not to rely on functions/views defined in `hdb_catalog` in the light of https://github.com/hasura/graphql-engine/issues/3533. The function body can be inlined in the statement itself?

During this refactor, I was also hoping that we would end up with simplified logic and fix some issues: 
### Better error messages
It would be good to have the object on which the constraint failed in the returned error message instead of the current "check constraint failed" (maybe something like "the following object '{"location":"blr"}' does not satisfy the check constraint"). This is particularly useful when you are inserting more than one objects or when inserting related data (through relationships). 

### Different code paths based on relationships in selection set
Assume the following schema:
```
user:
- id
- name
- addresses (array relationship to address table)
address:
- id
- user_id 
- location 
- user (object relationship to user table)
```
Let's say we issued a mutation such as this:
```graphql
mutation insert_address {
  insert_address(objects: {location: "blr"}) {
    affected_rows
    returning {
      user {
        addresses {
          id
          location
      }
    }
  }
}
```
Now, ideally you would expect to generate SQL such as this:
```
with "insert_result" AS ( INSERT INTO .. )
select .. from "insert_result"
```
However, this doesn't work as expected (the address inserted as a part of this mutation will not be returned in the `addresses` relationship) because of how CTEs [behave](https://www.postgresql.org/docs/current/queries-with.html) :
> All the statements are executed with the same snapshot (see Chapter 13), so they cannot “see” one another's effects on the target tables

So, whenever there are relationships in `returning`'s selection set, what we do is something different:
```sql
BEGIN
inserted rows <- INSERT INTO ...
WITH (inserted rows) AS "insert_result" SELECT .. from "insert_result"
COMMIT
```
This also applies in case of deletes and updates. I was wondering if having the same code path would reduce some of the issues that we have been seeing with inserts (https://github.com/hasura/graphql-engine/issues/3148, https://github.com/hasura/graphql-engine/pull/3375, https://github.com/hasura/graphql-engine/issues/3609). 

I have to admit, I'm not 100% familiar with the current insert mutations implementation so I do not know the implication of the changes that I'm proposing. @rakeshkky thoughts?

**Comment by paf31** [comment]
@0x777 I'm not sure how I would inline the function body into the call site since it contains a `RAISE` statement.

I'm all for consolidating the different code paths. To be honest, I found the current way quite confusing, and I was trying to preserve the existing behavior.

File: server/src-lib/Hasura/SQL/DML.hs [hunk 1]
```diff
@@ -283,7 +283,8 @@ data SQLExp
   | SELit !T.Text
   | SEUnsafe !T.Text
   | SESelect !Select
-  | SEStar
+  | SEStar !(Maybe Qual)
```
Comment (paf31) [inline server/src-lib/Hasura/SQL/DML.hs]: This isn't needed any more, but it seems like a useful addition anyway. Keep or revert?

Comment (lexi-lambda) [reply server/src-lib/Hasura/SQL/DML.hs]: Seems fine to keep it to me.

File: docs/graphql/manual/deployment/downgrading.rst [hunk 2]
```diff
@@ -103,6 +103,12 @@ You can downgrade the catalogue from a particular version to its previous versio
   :backlinks: none
   :depth: 1
   :local:
+  
+From 29 to 28
```
Comment (shahidhk) [inline docs/graphql/manual/deployment/downgrading.rst]: Catalog version as per this PR is 30

Shouldn't we have 30 to 29 downgrade instructions?

Comment (paf31) [reply docs/graphql/manual/deployment/downgrading.rst]: Yes, thanks. I forgot to change this when I merged master (which bumped from 29 to 30).

**Pull request merged**